packages/compose/src/higher-order/with-instance-id/index.tsx

```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentType } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import createHigherOrderComponent from '../../utils/create-higher-order-component';
+// eslint-disable-next-line no-duplicate-imports
+import type { PropRemovingHigherOrderComponent } from '../../utils/create-higher-order-component';
+import useInstanceId from '../../hooks/use-instance-id';
+
+/**
+ * A Higher Order Component used to be provide a unique instance ID by
+ * component.
+ */
+const withInstanceId: PropRemovingHigherOrderComponent< {
+	instanceId: string | number;
+} > = createHigherOrderComponent(
```

Comment:
Does it make sense to extract a `InstanceIdProps` or even export it? The type definition is used twice in the code.

Reply:
I originally did that but it made the extracted documentation worse because it made the type `PropRemovingHigherOrderComponent< InstanceIdProps >` instead of listing the props that are removed. Do you think it's still worth aliasing the type in that case?

Reply:
I would personally do it and document the `InstanceIdProps` type. But I don't mind either way.

To make documentation clearer, I would also rename the HOC type to `PropInjectingHigherOrderComponent`. Before it's too late after the NPM package is published and backward compatibility becomes an issue.

Reply:
FWIW we don't fully export the type from the package so there's no concern about change it immediately, but I do like that name better.